lib/Sema/ConstraintSystem.cpp

```diff
@@ -3559,6 +3559,11 @@ void constraints::simplifyLocator(ASTNode &anchor,
           continue;
         }
       }
+
+      if (isExpr<TypeExpr>(anchor)) {
+        path = path.slice(1);
```

Comment:
@theblixguy This is more a question, but should we update the anchor as the typeExpr too in this case? e.g. what is done for `ConstraintLocator::ConstructorMember`.

Reply:
Hmm, I suppose we can do it. ApplyArgument would handle it but I’m not sure if there is a scenario where ApplyArgToParam is not preceded by ApplyArgument path element.

Reply:
The problem here is related to the fact that subscript expression `viewControllers?[Int` doesn't have either `ParenExpr` nor `TupleExpr` wrapping an argument list, otherwise existing logic would have handled it correct. I don't think we need to check for `TypeExpr` here but instead when we see `ApplyArgument` with `SubscriptExpr` anchor we need to create a fake `ParenExpr` and set it as `anchor` for cases like this.

Reply:
Hmm okay sure, I'll have to plumb an `ASTContext` down to `simplifyLocator` for that.

Reply:
Seems rather tricky because there are callers where we don't have access to an `ASTContext`, I could make it optional though.

Reply:
If I remember correct you can get ASTContext from `TypeBase`.

Reply:
Oh absolutely, thanks for the reminder!

Reply:
Hmm unfortunately the expression doesn't have a type :/

Reply:
Okay, that's unfortunate... I wonder what it would take a fix this problem in a different way - make sure that even a single unlabeled argument associated with `SubscriptExpr` would have a wrapping paren around it to make what the rest of the calls look like...

Reply:
Maybe we can add one implicitly when we create a subscript expression?

Reply:
The first option is what I was thinking about, I don't know how much impact would that have though...

Reply:
It would be unfortunate to add a field to an expression which would be only used in one very particular situation...

Reply:
Let's try this one - when `ApplyArgument` has a `SubscriptExpr` anchor, check whether there is tuple or paren - if not and next element is `ApplyArgToParam` slice two elements and return argument.

Reply:
That's possibly the most narrow fix we could do and cherry-pick to 5.3

Reply:
Yeah, well I was thinking if we could change `getIndex` to look through the `ParenExpr` but add another method which would preserve it, but again that would be just like adding ASTContext for one-time use.

Another alternative might be to manually scan the codebase for each call to `getIndex` and look through the parens

Reply:
Okay let me try that!

Reply:
I think we need to scope this down to `simplifyLocator` so it's eligible for 5.3.

Reply:
Okay, I have done that now!